AWS: Delay all AWS calls when we observe RequestLimitExceeded errors

justinsb opened the pull request:
This applies a cross-request time delay when we observe
RequestLimitExceeded errors, unlike the default library behaviour which
only applies a _per-request_ backoff.

Issue #12121

justinsb committed:
AWS: Delay all AWS calls when we observe RequestLimitExceeded errors

This applies a cross-request time delay when we observe
RequestLimitExceeded errors, unlike the default library behaviour which
only applies a *per-request* backoff.

Issue #12121

justinsb commented:
Currently untested, but even more importantly there are a lot of heuristics here that I totally made up.  If anyone knows of the "correct" approach, that would be appreciated.

justinsb commented:
I think we should get something like this into 1.2, but I am really unsure what that something should be.  @danielschonfeld and I are fixing lots of individual causes (cf #19874) but there will always be a "next" bottleneck. @brendandburns or @davidopp any thoughts?

davidopp commented:
@justinsb at the rate things are going, I don't think anyone on our side is going to have time to give this significant thought before 1.2. (I hope @brendandburns will prove me wrong by jumping in here, but I think it is unlikely.) If you're satisfied that this PR solves the problem, let me know and I'll LGTM it.

davidopp commented:
ping @justinsb

davidopp commented:
ping

jtblin commented:
@justinsb here are my 2 cents about API limits and retries:

* API requests are distributed across nodes and distributed across processes on each node, per process logic seems to have limited benefits. I don't know if distributing this state in etcd is possible and scalable but it feels that it should be global to the cluster if possible.
* AWS API limits are per [category of requests](http://docs.aws.amazon.com/AWSEC2/latest/APIReference/query-api-troubleshooting.html) e.g. `ec2.DescribeInstances` and `ec2. DescribeVolumes` , for elb, for routes, etc. I am not sure of the exact granularity (i.e. each request has its own limit?) but imho custom logic should take that into account instead of a global delay otherwise we are going to end-up delaying requests for no reason.
* Waiting to observe the `RequestLimitExceeded` error might be too late so we should implement a rate limiter before hitting the limits.
* The aws-sdk have an exponential backoff retry logic (`maxRetries = 3`). I believe each error counts toward the limits so it may compound the problems (I am not sure about this).
* It is possible to implement custom retry logic by implementing the aws [request.Retryer](https://github.com/aws/aws-sdk-go/blob/master/aws/request/retryer.go) interface, see https://github.com/aws/aws-sdk-go/blob/master/aws/client/default_retryer.go. Maybe a first step would be to change the default retry backoff to be at least 1 second (AWS limits are 5/second and I think with the default the 3 retries will be in less than a second).

With all that said, your approach is quite simple and elegant in the end and it might do the trick without having to build something too sophisticated.

justinsb commented:
@jtblin the big problems is that we don't know the rate limits; and they might actually vary based on AWS load (!).  So even if we could coordinate calls cluster-wide, we have no way to avoid hitting the limits.

Given that, I'm inclined to just put something quick-and-simple into 1.2.

The Retryer interface is attractive, except that it only happens after we've hit an error.  What we want to do is throttle all requests on all threads when we think we're over the limit.

justinsb commented:
I'm almost inclined to just put this in as-is, despite the heuristics being basically arbitrary, on the grounds that a cluster that is hitting rate limits is in trouble anyway, and this is unlikely to make things worse...

jtblin commented:
@justinsb 

> the big problem is that we don't know the rate limits; and they might actually vary based on AWS load (!). So even if we could coordinate calls cluster-wide, we have no way to avoid hitting the limits.

The limits are 5 requests per second unless they have been raised, but yes there is no way to check what the current usage is as there can be other calls in the account outside of the k8s cluster and it's unclear when the timer starts / stops. Anyhow I wasn't suggesting to track the _number of API requests_ cluster-wide, but the _delay_. And instead of doing it globally for all requests, doing it per API category. But this would be much more complicated and may end up DoS'ing `etcd` so not sure it's worth it. 

> The Retryer interface is attractive, except that it only happens after we've hit an error. What we want to do is throttle all requests on all threads when we think we're over the limit.

In your current PR, you are also triggering the delay after we've hit the error but too late to recover from it iiuc (`AfterRetry`). I think what's attractive with this interface is that it is the AWS native way, it's specific to the current API call (so other API type calls are not delayed arbitrarily) and it allows to react on the first `RequestLimitExceeded` occurrence instead of after 3 occurrences (default for `maxRetries`) which gives more chance to succeed the request by waiting more than 1 second before retrying. 

> I'm almost inclined to just put this in as-is, despite the heuristics being basically arbitrary, on the grounds that a cluster that is hitting rate limits is in trouble anyway, and this is unlikely to make things worse...

SGTM. I'll try and send a simple example using the `Retryer` interface for comparison.

justinsb commented:
I think we should put this in to 1.2.  I know it's late, but I think the biggest uncertainty here is as to the parameters, not as to whether we need this.  We can tweak those parameters easily in patches and/or the final 1.2 process.

In general though, we should never hit this code.  If we hit it, it is because something has gone seriously wrong elsewhere.  Currently our behaviour when we hit API rate limits is pretty bad - we typically retry right away.  We're working on fixing these, but I'm sure we haven't got them all and this is a pretty vital safety belt therefore.

@jtblin also has a similar PR #21137, but this one hits the first request, not just retries which I think is important.

In short: yes it's risky, but we should never hit this code unless something has already gone badly wrong.  The parameters are likely wrong, but we're never going to be able to tune them unless we put it into a release.

cc @brendandburns

justinsb commented:
Reassigned to brendan... I don't think this is perfect, but I think it's better than the snowballing problems that happen when we hit the rate limits.

bprashanth commented on pkg/cloudprovider/providers/aws/retry_handler.go:
```diff
@@ -0,0 +1,137 @@
+	if delay > 0 {
+		glog.Warningf("Inserting delay before AWS request (%s) to avoid RequestLimitExceeded: %s",
+			describeRequest(r), delay.String())
+		r.Config.SleepDelay(delay)
```
I'm guessing this will just block the caller as opposed to returning a RetryRequest error? (and I guess that's ok from the client perspective). Can you please add a comment. 

I guess it would be nice if you returned the error so the caller can, for example, submit the request to another region (these are per region right?). Just a thought.

justinsb replied:
Will add a comment.

This is intended as a last resort; it will likely end up with the k8s cluster becoming slow/unresponsive.  But we have to protect the AWS account; once we start hitting rate limits both k8s and anything else running on the account cease to be operation.

This has to happen in conjunction with retry policies in controllers, for example.  But we know we won't find all the error cases, and so this is our last line of defense.

bprashanth commented on pkg/cloudprovider/providers/aws/retry_handler.go:
```diff
@@ -0,0 +1,137 @@
+
+// Added to the AfterRetry chain; called after any error
+func (c *CrossRequestRetryDelay) AfterRetry(r *request.Request) {
+	if r.Error != nil {
```
nit: I prefer to avoid the indent and just return if r.Error == nil

justinsb replied:
Agreed - thanks!

bprashanth commented on pkg/cloudprovider/providers/aws/retry_handler.go:
```diff
@@ -0,0 +1,137 @@
+		if awsError, ok := r.Error.(awserr.Error); ok {
+			if awsError.Code() == "RequestLimitExceeded" {
+				c.mutex.Lock()
+				c.countErrorsRequestLimit += 1.0
```
would be nice if you could encapsulate this in a Backoffer interface that just has methods like RecordError(err), GetDelay(). RecordError is responsible for doing the whole Code() == bit, or whatever the cloudprovider specific thing is. Feel free to defer this to another pr. Ideally we'd initialize the struct that implements this interface with variables like min, max, jitter etc.

justinsb replied:
This is a great idea actually as then I think I could implement at least some form of test... I'll do that

bprashanth commented on pkg/cloudprovider/providers/aws/retry_handler.go:
```diff
@@ -0,0 +1,137 @@
+	c.mutex.Lock()
+
+	// Apply exponential decay to the counters
+	timeDeltaSeconds := now.Unix() - c.lastDecay
```
suggest using util.Clock so we can fake it out? https://github.com/kubernetes/kubernetes/blob/master/pkg/util/clock.go#L26. And if you do that please either write or add a todo for a simple unittest.

justinsb replied:
I added a set of simple unit tests that test the "macro" behaviour e.g. no errors => no delay; errors => delay with no flapping; errors go close to max eventually; recovers from errors eventually.

There wasn't a need to mock the clock, because computeDelay takes now as an argument anyway.

bprashanth commented on pkg/cloudprovider/providers/aws/retry_handler.go:
```diff
@@ -0,0 +1,137 @@
+		// Avoid tiny residuals & rounding errors
+		errorFraction = c.countErrorsRequestLimit / c.countRequests
+	}
+	c.mutex.Unlock()
```
is there a reason to not simply defer?

justinsb replied:
I think it was an artifact of when I used to sleep in the function.  Moving to defer!

bprashanth commented on pkg/cloudprovider/providers/aws/retry_handler.go:
```diff
@@ -0,0 +1,137 @@
+	mutex sync.Mutex
+
+	// We count all requests and the number of requests which hit a
+	// RequestLimit We only really care about 'recent' requests, so we
```
did you mean to put a full stop before "We"?

justinsb replied:
Thanks (I reflowed the text, I think)

justinsb commented:
Thanks for the great comments @bprashanth - I think I addressed all yours comments, including adding some tests - thanks for the hint about splitting out the code & mocking time.  The refactor touched most of the lines, so I squashed the commits.

bprashanth commented on pkg/cloudprovider/providers/aws/retry_handler.go:
```diff
@@ -0,0 +1,154 @@
+// Create a new CrossRequestRetryDelay
+func NewCrossRequestRetryDelay() *CrossRequestRetryDelay {
+	c := &CrossRequestRetryDelay{}
+	c.backoff.init()
```
Why not just init it with all the delay/interval stuff instead of making it reach out to globals? or add a constructor like NewBackoffer(decay, maxDelay...)

justinsb replied:
I'll have to duplicate it in the tests.  I'll do it though...

bprashanth commented on pkg/cloudprovider/providers/aws/retry_handler_test.go:
```diff
@@ -0,0 +1,135 @@
+	b := &Backoff{}
+	b.init()
+
+	now := time.Now()
```
This is the typical use case of a fake clock, so you don't need to depend on the system clock in a unittest. Eg: https://github.com/kubernetes/kubernetes/blob/42e2ff94c15c01686473665d024a4bd1a5bf8916/pkg/client/cache/expiration_cache_test.go#L115

justinsb replied:
I only use time.Now because it's the simplest way to build a time value :-)  Then ComputeDelayForRequest naturally takes a timestamp, because of the need to set "r.Time = now" to avoid clock skew problems with the aws-sdk.

That said, I think it's a total wash - happy to switch to a fake clock if that's clearer...

bprashanth replied:
Upto you, yeah the real danger case is subtracting time values in a unit test because the system clock can just randomly reset causing flake. As long as we're not hitting that it's fine.

bprashanth commented:
LGTM but for the nit about not grabbing the timing globals in the Backoff struct

bprashanth commented:
Your e2e failure is https://github.com/kubernetes/kubernetes/issues/21138, if you need a new run and decide to not upload a commit with the New/Init(params...) suggestion

justinsb committed:
AWS: Pass globals into Backoff struct

Thanks for the suggestion bprashanth!

justinsb commented:
Thanks @bprashanth - I added the init params as you suggested.  Hopefully I'll get lucky on this e2e run!

bprashanth commented:
LGTM thanks
